superduperdb/cluster/task_workflow.py

```diff
@@ -1,66 +1,73 @@
-import datetime
 import networkx
+import copy
+import datetime
+import dataclasses as dc
+import functools
+import typing as t
 import uuid
 
 from superduperdb.cluster.function_job import function_job
 from superduperdb.cluster.job_submission import dask_client
 
+Self = t.Any
+
 
+@dc.dataclass
 class TaskWorkflow:
-    def __init__(self, database):
-        self.G = networkx.DiGraph()
-        self._path_lengths = None
-        self.database = database
+    database: t.Any
+    G: networkx.DiGraph = dc.field(default_factory=networkx.DiGraph)
 
-    def add_edge(self, node1, node2):
+    def add_edge(self, node1, node2) -> None:
         self.G.add_edge(node1, node2)
 
-    def add_node(self, node, data=None):
+    def add_node(self, node, data=None) -> None:
         data = data or {}
         self.G.add_node(node)
-        for k, v in data.items():
-            self.G.nodes[node][k] = v
+        self.G.nodes[node].update(data)
 
-    def __mul__(self, other):
+    def asjson(self) -> t.Dict:
```

Comment:
```suggestion
    def asjson(self) -> t.Dict[t.Any, t.Any]:
```
It is better to be explicit (and the issue I'm working on will also raise an error if we are not)

Reply:
`t.Dict[t.Any, t.Any]` identifies exactly the same type as `t.Dict`, so why the extra bulk?

Does `mypy` really distinguish between these two, and if so, why?

Reply:
(Also, you can say `isinstance(d, t.Dict)` but not `isinstance(d, t.Dict[t.Any, t.Any])`.)

Reply:
I went through the networkx code and docs and it isn't clear what the keys are, but if it's going to really be `JSON`, then the keys must be `str`, so I changed it to `t.Dict[str, t.Any]`

Reply:
Sorry for this. It was my mistake.

The `[str, t.Any]` specification was not needed in fact.